### DISCUSSION TITLE
`mypy`/`pyright` differing behaviour unifying `Callable[[VarArg(Hashable),KwArg(Hashable)], _KT]`

### DISCUSSION THREAD

**Discussion by felebsp** [python/typing#1633] [Q&A]
While trying to create type stubs for `asyncache`, I came across this difference in behaviour between `mypy` and `pyright`.

The following code is accepted in the [pyright playground](https://pyright-play.net/?code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgAkCcALY5AKFEilQE8L0sAczKVqtAArhUYOPAA0UACp8kANQIgFAWQCuqVki0EK-IWzYATJMCgtmAayQ8AFACoNgnAC56jFkWQFV1d7AHcPb19mAwBKKABaAD5uHQpkAG0GaICkBTx8gF0fNihSqBAkVB0QLBS0pGcCKGBqKCaBNpBPGPMAfQAlRV64KABeJRV1EGcAcgGhuBmFOAA3DXQCLFRRxRAdJB62GHhGHCgAYQIYJiQDADEdLDRMLGdJMGlZdPnhsAKYrwlMoAARw%2BgwMAglSYYAsQNKVhsvWGBEQyLcER8WX8gSgwTCmKiOIOCWSPzggLKVKg%2BTw8OpbF6AGlFGMJhQ1BpZszFEtYGA1iANlsdnsDozWeNlByptzeYdEdwkGDesBHs56TArjcfLp9DkjCYBIJviyFL1FAU2QBvAC%2BcnpdiYaqwPku11uOQeTww2FNlrZTscPAdVPcXUi2IMbDiSSgFspZQqVRqsG1SHSTpdjQjMQK9LYqGVqDd6fuj2e2EDfmDbCAA)

```python
from collections.abc import Hashable
from typing import Protocol, TypeVar, MutableMapping

def hashkey(*args: Hashable, **kwargs: Hashable) -> tuple[Hashable, ...]: 
    return tuple(a for a in args)

_RT_co = TypeVar('_RT_co', covariant=True)

class CacheableFunction(Protocol[_RT_co]):
    @staticmethod
    def __call__(*args: Hashable, **kwargs: Hashable) -> _RT_co:
        ...
        
_KT = TypeVar('_KT', covariant=True)
_T = TypeVar('_T')

def test_fun(
    cache: MutableMapping[_KT, _T] = {},
    hashfun: CacheableFunction[_KT] = hashkey,
    *args: Hashable
) -> _T:
    return cache[hashfun(args)]
    
test: CacheableFunction = hashkey
```
However, [mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=5ebbc22f584bd78cb574977baa26739a) gives the error 
```
Incompatible default for argument "hashfun" (default has type "Callable[[VarArg(Hashable), KwArg(Hashable)], tuple[Hashable, ...]]", argument has type "CacheableFunction[_KT]")  [assignment]
main.py:19: note: "CacheableFunction[_KT].__call__" has type "Callable[[VarArg(Hashable), KwArg(Hashable)], _KT]"
```

Is this a bug? If so, of which type checker?

**Comment by erictraut** [comment]
This looks like a bug in mypy. If you change the last line to `test: CacheableFunction[tuple[Hashable, ...]] = hashkey`, you can see that mypy accepts this assignment without error.

**Comment by Daverball** [comment]
This error makes sense to me, since even though `CacheableFunction` is covariant the key type for `Mapping` isn't, so it's technically not type safe to accept `hashkey` for `CachableFunction[_KT]`, since `_KT` could be something other than `tuple[Hashable, ...]` if you supply your own mapping without also supplying a `hashfun`.

You can add some overloads to deal with this:

```python
@overload
def test_fun(
    cache: MutableMapping[tuple[Hashable, ...], _T] = {},
    hashfun: CacheableFunction[tuple[Hashable, ...]] = hashkey,
    *args: Hashable
) -> _T: ...


@overload
def test_fun(
    cache: MutableMapping[_KT, _T] ,
    hashfun: CacheableFunction[_KT],
    *args: Hashable
) -> _T: ...


def test_fun(
    cache: MutableMapping[Any, _T] = {},
    hashfun: CacheableFunction[Any] = hashkey,
    *args: Hashable
) -> _T:
```

As a side note, I don't think `Hashable` is very useful for type checking since types that fulfill the `Protocol` can still be unhashable such as a `tuple` containing an unhashable object. I've found that it's usually not worth the hassle and it's better to just go with `object` or in this case probably `Any`, since arguments are contravariant.

**Reply by erictraut** [reply]
@Daverball, I don't understand your analysis.

I don't think mypy's error has anything to do with `Mapping`, `Hashable` or covariance. It appears to be a more fundamental problem. You can simplify the sample to the follow, and mypy still emits an error.

```python
from typing import Protocol, TypeVar

def hashkey() -> None:
    pass

_RT_co = TypeVar("_RT_co", covariant=True)

class CacheableFunction(Protocol[_RT_co]):
    @staticmethod
    def __call__() -> _RT_co: ...

_KT = TypeVar("_KT")

def test_fun(hashfun: CacheableFunction[_KT] = hashkey) -> None:
    pass
```

**Reply by Daverball** [reply]
That's fair, I've come across this myself in the past and it seemed reasonable to me for mypy to reject it. If it still breaks once you set a default for `_KT` using PEP696 (and support is fully implemented, which seems to be coming along nicely, but isn't far along yet to support this case), then I would consider it a bug too. Without PEP696 it doesn't even really matter whether it's a bug or intended behavior, since it's better to write the overloads anyways, if you want proper type safety.

Although you could certainly argue that you shouldn't have to write `Any` in the implementation signature and the fact that you have to is a bug. Right now I'd call that behavior a happy accident, but it's a good test case for finishing PEP696 support, since the default value should be compatible with the type param's default type.

**Reply by Daverball** [reply]
You can also get the same error if you assign any concrete type as a default for a parameter with a generic:

```python
default_cache: MutableMapping[tuple[Hashable, ...], Any] = {}

def test_fun(
    cache: MutableMapping[_KT, _T] = default_cache,
    hashfun: CacheableFunction[_KT] = hashkey,
    *args: Hashable
) -> _T:
```

This will now also result in an invalid default value error for the `cache` parameter as well. I think in either case it's invalid to accept a concrete bound type other than `Any` because once you fix `_KT` you can construct examples where the default is no longer compatible, regardless of whether the type parameter is covariant, contravariant or invariant, apart from a few corner cases like `object` and `Never` in the case of co-/contravariance.

And you can actually go one step further and  you also get the same behavior when the argument is annotated as `_T`, then providing any concrete value with a type other than `_T` or `Any` is also invalid.

```python
def test_fun(value: _T = 5) -> None: ...  # error
```

**Reply by Daverball** [reply]
Whether or not this should be allowed is definitely up for debate, especially in the case of the last example. I think prohibiting it can help prevent a common class of mistakes and help the user recognize that type vars may conflict with the default value when calling the function with a certain subset of arguments, since the default values are not part of the signature and thus not considered when checking calls.

But the restriction could be relaxed with type var defaults, as long as the rule for default values is that their version of the type parameter is bound to the default, so if another argument that was provided binds that same type parameter to a different type that is incompatible with the default, then that should be an error, same as if you had explicitly spelled out the individual overloads.

**Reply by erictraut** [reply]
> once you fix `_KT` you can construct examples where the default is no longer compatible

`_KT` doesn't get a value until the generic function is invoked in a call expression. If there's any error here, I would expect it to be reported at the call site, not the function definition site. That's how pyright implements this support. Mypy could do the same.

> since the default values are not part of the signature and thus not considered when checking calls

Yes, if the default value is not supplied (e.g. is specified as `...`, as it sometimes is within stub files), then a type checker would not be able to verify compatibility at the call site. However, in real code (non-stubs) the default argument is always provided and can be used. There is also a trend to supply default argument values in stubs as well because they are valuable for language servers.

There are multiple active issues in the mypy issue tracker related to this use case. The earliest one is [this one](https://github.com/python/mypy/issues/3737), which has been open for 7 years and has more than 80 upvotes. Regardless of whether mypy maintainers consider this a feature request or a bug, I think this is an issue with mypy specifically. I don't think there's a strong justification in the type system for generating an error here.

**Reply by Daverball** [reply]
I think both ways are valid approaches, if the default values aren't considered at the call-site then we had better made sure the definition site is strict enough to catch potential errors involving default values, but when they are considered, you can relax this restriction. So it's more of an implementation detail for how you prevent this error and less about it being faulty behavior. The spec could rectify this by specifying how default values should be treated at definition and call time, including a recommendation when the default value has been ellided (this can also happen in overloads without error, not just in stubs, although in that case the value should usually be inferable from the implementation's signature).

FWIW I do think pyright's approach is better when we know the default value, because it ensures safety even if we made a consistency mistake in the definition that wasn't caught.

**Reply by Daverball** [reply]
Thinking on it a bit further, there is one thing where pyright's approach is inherently less type safe, unless you apply complex logic and that is when you cast a callable with optional parameters to another compatible callable (based solely on the argument types) where some of those parameters are missing. So now you are stripping the required information that allowed to ensure type consistency at call-time while simultaneously encouraging a scenario where type consistency is violated, so there would need to be some logic at the cast-site to detect safety holes, so mypy's approach is more secure there, since it tries to ensure that you can't even write a definition with this class of problems, which is less ergonomic and more annoying to use, but I can also see why you would choose to nevertheless go this way, since it avoids introducing safety holes or complex logic at type analysis boundaries.

**Reply by erictraut** [reply]
I'm not sure what you mean by "cast" in this context. Are you referring to a call to `typing.cast`? That's inherently unsafe since it subverts all type checks. If that's not what you mean, perhaps an example would help clarify.

**Reply by Daverball** [reply]
I mean an implicit cast, i.e. anywhere you would perform a subtype or supertype check for compatibility between two types. Which mypy still performs in the case of `typing.cast`, it doesn't allow arbitrary incompatible casts without ignoring a type error.

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.12&code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGEDFjkBYAKHYBMlgCBXeKgD6MAjAAWSAFxRO6NAG0cqEABoyWVAF0oAXigBvAL5ceUYGDAKAKloAUADxnX1ATxlzFytVFt7ZPPyCImKSAJRQALQAfL4yeAnspsBQqAQA1kg4Qujcmuiorjb2wHxYMDIMTETICsXqtuoAbox80r4RMVAActjS7FCD5mUwdi3wbWFJbGmZ2blI%2BYV2FmDqAERlOATASOsRUADEUFhgUEgg4CBAA)

```python
from collections.abc import Callable

default_cache: dict[str, int] = {}
def foo[T](x: T, y: dict[str, T] = default_cache) -> T: ...

def takes_identity[T](func: Callable[[T], T], value: T) -> None:
    func(value)

takes_identity(foo, "unsafe")  # no error
```
